### ISSUE TITLE
Generate Urls outside of a request (cron task)

### ISSUE THREAD

**Issue by do0om** [aspnet/Mvc#5164]
Hello,

I am using [hangfire ](http://hangfire.io/) for a cron task sending email reminders in my asp.net core app.
In this method I need to generate urls that will be embedded into emails.

I wish I could have access to a IUrlHelper.
But according to #5071, this is not possible, because we need an httpContext that is not available outside of a request. I tried to get a IUrlHelperFactory through dependency injection and I got one but it could not produce a valid IUrlHelper

Then I had the idea of using an IRouter instance like [here](https://docs.asp.net/en/latest/fundamentals/routing.html#url-generation-reference) but I couldn't get
access to one through dependency injection.

This is what my cron service  looks like 

```
    public class CronService
    {
 
        public CronService( /* IUrlHelperFactory urlHelperFactory*/)
        {
           
        }

        public void SendMailTask()
        {
            IUrlHelper url = GetUrlHelper();//pseudo code : I can't get url helper here
            string loginUrl = url.Action("Login", "Account");

            SendMail(loginUrl);
        }
    }
```
It has been registered in Startup.cs

And my hangfire registering call in Startup.cs is

 ` RecurringJob.AddOrUpdate<CronService>(cronService => cronService.SendMailTask(), Cron.Minutely);
`

What can I do?

Thanks

**Comment by Eilon** [comment]
@do0om if you got an `IUrlHelperFactory` then I'd say you're almost there. You'll need to manufacture a "fake" `ActionContext` that has just enough info on it for routing to generally work (which isn't much). Routing doesn't use a whole lot of info from the context in most cases, so having most of the context info just return null/empty/default, will probably work fine.

**Comment by do0om** [comment]
@Eilon  Thanks for the answer and sorry for my late reply.
I tried several configurations to get a “fake” `ActionContext` as you said but it could not make it work. Here are my attempts:
 ```
IUrlHelper helper = urlHelperFactory.GetUrlHelper(new ActionContext());
//not working : ArgumentException  The 'HttpContext' property of 'ActionContext' 
//must not be null.

ActionContext actionContext = actionContextAccessor.ActionContext;
//=> is null, but that's not surprising : I am outside of an action

//trying to get an url helper with default values
//helper is returned but it's not working
IUrlHelper helper1=urlHelperFactory.GetUrlHelper(new ActionContext(
                new DefaultHttpContext(),
                new RouteData(),
                new ActionDescriptor() {} ));

string url=helper1.Action("Index", "Home");
//=> System.ArgumentOutOfRangeException
```
I looked a bit into the source code and I see : 
` protected IRouter Router => ActionContext.RouteData.Routers[0];`
 again, I think my problem is in getting this `IRouter` instance

**Comment by Eilon** [comment]
@rynowak thoughts on how to fake enough data to use this API outside of a web request?

**Comment by rynowak** [comment]
The route collection is data. You need to get the route collection and then put it in `.Routers[0]` if you're using `IUrlHelper` or if you have the route collection, you can call `GetVirtualPath` directly on it.

One approach would be something like this:


```
// In Configure(...)

IRouter routeCollection = null;
app.UseMvc(routes =>
{
    // Add routes
    ...

    routeCollection = routes.Build();
}
```

**Comment by do0om** [comment]
This way I can access `IRouter` from `Startup Configure` method but how can I access it in my service?

**Comment by Eilon** [comment]
@do0om I think the idea is to store the `IRouter` wherever you want such that you can get to it later. E.g. store in a `static` field somewhere, store it in DI itself via your own custom service, etc.

**Comment by do0om** [comment]
Aren't static variables "dangerous" in asp.net?
For the DI solution, this looks cleaner but how can I add a service to DI in configure method (since this is where I build the route)?

**Comment by Eilon** [comment]
Static variables are not intrinsically dangerous in ASP.NET, not at all. It really depends on the usage and ensuring things like thread safety, race conditions, etc. But since this is startup code that runs only once, setting a static is fine from that perspective. After that it's read-only. The other potential concern is code cleanliness, but that doesn't automatically rule out all usage of statics either. Good code can be written with or without statics 😄

But really the key is of course just that you need *some* way to store the data and get it back "later" from your URL generation logic.

**Comment by do0om** [comment]
Thanks @Eilon ,I will use static then ;)

**Comment by uhfath** [comment]
Is there anything like this but for static files?
We also have a cron job in the project that sends emails, but each email is a razor rendered view, which contains some links to static files. And these links should be absolute.
The problem is that IUrlHelper generates URL only for actions and not for static files (or just arbitrary virtual->absolute conversion).
Tried using IHttpContextAccessor, but it's HttpContext.Request comes from DefaultHttpContext at the moment of generating email and so there is no Host or proper Request info in it.
Is there any way around this?

**Comment by mnmr** [comment]
I think it's insane that UrlHelper depends on so much request-infrastructure: it is just bad design. For instance, to use UrlHelper from a SignalR Hub (which has access to none of this stuff), you need to manually hack stuff together ([as done here](https://stackoverflow.com/questions/46096068/asp-net-core-2-0-creating-urlhelper-without-request). 

But clearly, since it possible to have a working UrlHelper without a valid request, why even require it in the first place? If you have some obscure scenario where it's really needed, provide overloads that let you pass in a context, instead of forcing this garbage upon people as a default requirement.

**Comment by aventurella** [comment]
I have the same need. I need to get a Url outside of a request. The 

**Startup.cs**
```
static IRouter _routes;

public void Configure(IApplicationBuilder app){
   app.UseMvc(routes => {
        ...
        _routes = routes.Build();
    })

    var lifetime = app.ApplicationServices.GetService<IApplicationLifetime>();
    lifetime.ApplicationStarted.Register(( ) =>
    {
        
        var routeData = new RouteData();
        routeData.PushState(_routes, new RouteValueDictionary(), new RouteValueDictionary());

        var urlHelperFactory = app.ApplicationServices.GetRequiredService<IUrlHelperFactory>();
        IUrlHelper helper = urlHelperFactory.GetUrlHelper(
            new ActionContext(
                new DefaultHttpContext(),
                routeData,
                new ActionDescriptor() { })
        );
        
        // The following 2 both throw the Exeption:
        // System.ArgumentNullException: Value cannot be null.
        var url1 = helper.Action(action: "Index", controller: "Test", values: null);
        var url2 = helper.RouteUrl("test_route");
    });
}

**Comment by rynowak** [comment]
**Hi, it looks like you are posting on a closed issue/PR/commit!**

We're very likely to lose track of your bug/feedback/question unless you:
1. Open a new issue
2. Explain very clearly what you need help with
3. If you think you have found a bug, include detailed repro steps so that we can investigate the problem

Thanks!
